File: web/src/types/divination.ts
/**
 * 占卜相关类型定义
 * 
 * 包含占卜结果、会话、卦象信息等核心数据结构
 */

/**
 * 占卜状态枚举
 */
export type DivinationStatus = 'processing' | 'completed' | 'failed' | 'pending';

/**
 * 占卜版本类型
 */
export type DivinationVersion = 'CN' | 'Global' | 'TAROT';

/**
 * 事件类型
 */
export type EventType = 'decision' | 'career' | 'relationship' | 'fortune' | 'knowledge' | 'health' | 'wealth' | 'general';

/**
 * 占卜结果
 */
export interface DivinationResult {
  /** 会话ID */
  session_id: string;
  /** 占卜结果（吉/凶/平） */
  outcome?: string;
  /** 标题 */
  title?: string;
  /** 牌阵类型 */
  spread?: string;
  /** 塔罗牌抽取结果 */
  cards?: TarotCardDraw[];
  
  /** 结果摘要（50-100字） */
  summary: string;
  
  /** 详细内容（点击展开） */
  detail: string;
  
  /** 卦象信息（决策类） */
  hexagram_info?: HexagramInfo;
  /** 推荐列表（推荐类） */
  recommendations?: RecommendationItem[];
  /** 每日运势（运势类） */
  daily_fortune?: DailyFortuneInfo;
  /** 大衍筮法起卦过程（仅周易六爻） */
  yarrow_trace?: YarrowProcessTrace;
  
  /** 原始数据 */
  raw_data?: string;
  /** 场景建议 */
  scene_advice?: SceneAdviceItem[];
  /** 是否需要追问 */
  needs_follow_up: boolean;
  /** 追问问题 */
  follow_up_question?: FollowUpQuestion;
  /** 创建时间 */
  created_at: string;
  /** 状态 */
  status?: DivinationStatus;
}

/**
 * 卦象信息
 */
export interface HexagramInfo {
  /** 卦序号（1-64） */
  number: number;
  /** 卦名 */
  name: string;
  /** 上卦 */
  upper_trigram: string;
  /** 下卦 */
  lower_trigram: string;
  /** 吉凶 */
  outcome: string;
  /** 卦辞摘要 */
  summary: string;
  /** 五行 */
  wuxing: string;
  /** 变爻位置 */
  changing_lines?: number[];
  /** 六爻值（自下而上，6/7/8/9） */
  line_values?: number[];
}

/**
 * 大衍筮法：单次变的记录
 */
export interface YarrowChangeStep {
  /** 第几变（1-3） */
  step_index: number;
  /** 本变开始前蓍草数 */
  stalks_before: number;
  /** 左手蓍草数 */
  left_pile: number;
  /** 右手蓍草数（挂一前） */
  right_pile_before_hang: number;
  /** 挂一数量（固定为1） */
  right_hang_one: number;
  /** 右手蓍草数（挂一后） */
  right_pile_after_hang: number;
  /** 左手取四余数（0按4计） */
  left_remainder: number;
  /** 右手取四余数（0按4计） */
  right_remainder: number;
  /** 本次去除总数 */
  removed: number;
  /** 本变结束后蓍草数 */
  stalks_after: number;
}

/**
 * 大衍筮法：单爻生成过程
 */
export interface YarrowLineTrace {
  /** 爻位（1-6，自下而上） */
  line_index: number;
  /** 初始蓍草数（通常49） */
  initial_stalks: number;
  /** 三变过程 */
  changes: YarrowChangeStep[];
  /** 三变后剩余蓍草数 */
  final_stalks: number;
  /** 爻值（6/7/8/9） */
  line_value: number;
  /** 爻类型（老阴/少阳/少阴/老阳） */
  line_type: string;
  /** 是否变爻 */
  is_changing: boolean;
}

/**
 * 大衍筮法：完整起卦过程
 */
export interface YarrowProcessTrace {
  /** 起卦方法，固定为 "dayan_yarrow" */
  method: string;
  /** 总蓍草数（50） */
  total_stalks: number;
  /** 参与演算蓍草数（49） */
  effective_stalks: number;
  /** 六爻过程（自下而上） */
  lines: YarrowLineTrace[];
}

/**
 * 推荐项
 */
export interface RecommendationItem {
  /** 推荐内容 */
  content: string;
  /** 推荐理由 */
  reason: string;
}

/**
 * 每日运势信息
 */
export interface DailyFortuneInfo {
  /** 综合评分 */
  overall_score: number;
  /** 财运评分 */
  wealth_score: number;
  /** 事业评分 */
  career_score: number;
  /** 爱情评分 */
  love_score: number;
  /** 健康评分 */
  health_score: number;
  /** 运势内容 */
  content: string;
  
  /** 幸运颜色 */
  lucky_color: string;
  /** 幸运数字 */
  lucky_number: number;
  /** 幸运方位 */
  lucky_direction: string;
  /** 幸运时辰 */
  lucky_time: string;
  
  /** 宜（逗号分隔） */
  yi: string;
  /** 忌（逗号分隔） */
  ji: string;
  
  /** 节气 */
  solar_term: string;
  /** 节日 */
  festival: string;
}

/**
 * 场景建议项
 */
export interface SceneAdviceItem {
  /** 标题 */
  title: string;
  /** 内容 */
  content: string;
  /** 类型 */
  type: string;
}

/**
 * 追问问题
 */
export interface FollowUpQuestion {
  /** 问题ID */
  id: string;
  /** 会话ID */
  session_id: string;
  /** 问题内容 */
  question: string;
  /** 选项列表 */
  options?: string[];
}

/**
 * 塔罗牌抽取结果
 */
export interface TarotCardDraw {
  /** 牌名 */
  name: string;
  /** 英文名 */
  name_en?: string;
  /** 位置 */
  position: string;
  /** 是否逆位 */
  is_reversed: boolean;
}

/**
 * 创建占卜请求
 */
export interface CreateDivinationRequest {
  /** 用户ID */
  user_id: string;
  /** 问题 */
  question: string;
  /** 事件类型 */
  event_type?: EventType;
  /** 版本 */
  version?: DivinationVersion;
  /** 方位 */
  orientation?: string;
  /** 牌阵 */
  spread?: string;
  /** 额外上下文信息（用于传递洗牌/切牌交互等） */
  context?: any;
}

/**
 * 占卜会话
 */
export interface DivinationSession {
  /** 会话ID */
  id: string;
  /** 用户ID */
  user_id: string;
  /** 版本 */
  version: DivinationVersion;
  /** 问题 */
  question: string;
  /** 事件类型 */
  event_type: EventType;
  /** 方位 */
  orientation?: string;
  /** 牌阵 */
  spread?: string;
  /** 状态 */
  status: DivinationStatus;
  /** 追问次数 */
  follow_up_count: number;
  /** 创建时间 */
  created_at: string;
  /** 更新时间 */
  updated_at?: string;
  /** 占卜结果（可选） */
  outcome?: string;
}

/**
 * API 错误响应
 */
export interface DivinationError {
  /** 错误消息 */
  message: string;
  /** 错误代码 */
  code?: string;
  /** 详细信息 */
  detail?: string;
  /** HTTP 状态码 */
  statusCode?: number;
}

File: backend-python/app/schemas/divination.py
"""占卜相关的 Pydantic 模式"""

from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime


class HexagramInfo(BaseModel):
    """卦象信息"""
    number: int = Field(..., description="卦序号（1-64）")
    name: str = Field(..., description="卦名")
    upper_trigram: str = Field(..., description="上卦")
    lower_trigram: str = Field(..., description="下卦")
    outcome: str = Field(..., description="吉凶判断")
    summary: str = Field(..., description="卦辞摘要")
    wuxing: str = Field(..., description="五行")
    changing_lines: Optional[List[int]] = Field(None, description="变爻位置")
    line_values: Optional[List[int]] = Field(None, description="六爻值（自下而上，6/7/8/9）")


class TarotCard(BaseModel):
    """塔罗牌"""
    name: str = Field(..., description="牌名")
    name_en: Optional[str] = Field(None, description="英文名")
    position: str = Field(..., description="位置（如：过去、现在、未来）")
    is_reversed: bool = Field(False, description="是否逆位")
    meaning: Optional[str] = Field(None, description="牌义")


class RecommendationItem(BaseModel):
    """推荐项"""
    content: str = Field(..., description="推荐内容")
    reason: str = Field(..., description="推荐理由")


class DailyFortuneInfo(BaseModel):
    """每日运势信息"""
    overall_score: int = Field(..., ge=0, le=100, description="综合评分")
    wealth_score: int = Field(..., ge=0, le=100, description="财运评分")
    career_score: int = Field(..., ge=0, le=100, description="事业评分")
    love_score: int = Field(..., ge=0, le=100, description="感情评分")
    health_score: int = Field(..., ge=0, le=100, description="健康评分")
    content: str = Field(..., description="运势内容")
    lucky_color: str = Field(..., description="幸运色")
    lucky_number: int = Field(..., description="幸运数字")
    lucky_direction: str = Field(..., description="幸运方位")
    lucky_time: str = Field(..., description="幸运时辰")
    yi: str = Field("", description="宜（逗号分隔）")
    ji: str = Field("", description="忌（逗号分隔）")
    solar_term: str = Field("", description="节气")
    festival: str = Field("", description="节日")


class YarrowChangeStep(BaseModel):
    """大衍筮法单次变的记录"""
    step_index: int = Field(..., ge=1, le=3, description="第几变（1-3）")
    stalks_before: int = Field(..., description="本变开始前蓍草数")
    left_pile: int = Field(..., description="左手蓍草数")
    right_pile_before_hang: int = Field(..., description="右手蓍草数（挂一前）")
    right_hang_one: int = Field(..., description="挂一数量，固定为1")
    right_pile_after_hang: int = Field(..., description="右手蓍草数（挂一后）")
    left_remainder: int = Field(..., description="左手取四余数（0按4计）")
    right_remainder: int = Field(..., description="右手取四余数（0按4计）")
    removed: int = Field(..., description="本变去除总数")
    stalks_after: int = Field(..., description="本变结束后蓍草数")


class YarrowLineTrace(BaseModel):
    """单爻生成过程"""
    line_index: int = Field(..., ge=1, le=6, description="爻位（1-6，自下而上）")
    initial_stalks: int = Field(..., description="初始蓍草数，通常49")
    changes: List[YarrowChangeStep] = Field(default_factory=list, description="三变过程")
    final_stalks: int = Field(..., description="三变后剩余蓍草数")
    line_value: int = Field(..., description="爻值（6/7/8/9）")
    line_type: str = Field(..., description="爻类型（老阴/少阳/少阴/老阳）")
    is_changing: bool = Field(..., description="是否变爻")


class YarrowProcessTrace(BaseModel):
    """大衍筮法完整过程记录"""
    method: str = Field("dayan_yarrow", description="起卦方法")
    total_stalks: int = Field(50, description="总蓍草数")
    effective_stalks: int = Field(49, description="参与演算蓍草数")
    lines: List[YarrowLineTrace] = Field(default_factory=list, description="六爻过程（自下而上）")


class DivinationResult(BaseModel):
    """占卜结果"""
    session_id: str
    status: Optional[str] = Field(None, description="状态：processing/completed/failed")
    error_code: Optional[str] = Field(None, description="失败错误码")
    error_message: Optional[str] = Field(None, description="失败错误信息")
    retryable: Optional[bool] = Field(None, description="是否可重试")
    outcome: Optional[str] = None  # 吉/凶/平
    title: Optional[str] = None
    spread: Optional[str] = None
    cards: Optional[List[TarotCard]] = None
    
    # 结果内容（processing 状态时可为空）
    summary: Optional[str] = Field(None, description="简要结果")
    detail: Optional[str] = Field(None, description="详细解释")
    
    # 结构化数据
    hexagram_info: Optional[HexagramInfo] = None
    recommendations: Optional[List[RecommendationItem]] = None
    daily_fortune: Optional[DailyFortuneInfo] = None
    yarrow_trace: Optional[YarrowProcessTrace] = Field(None, description="大衍筮法过程记录")
    
    # 问题分析信息
    question_type: Optional[str] = None
    question_intent: Optional[str] = None
    
    needs_follow_up: bool = False
    created_at: datetime
    
    class Config:
        json_schema_extra = {
            "example": {
                "session_id": "uuid-here",
                "status": "completed",
                "outcome": "吉",
                "summary": "此卦大吉，诸事顺利",
                "detail": "详细的卦象解释...",
                "hexagram_info": {
                    "number": 1,
                    "name": "乾",
                    "upper_trigram": "乾",
                    "lower_trigram": "乾",
                    "outcome": "吉",
                    "summary": "元亨利贞",
                    "wuxing": "金",
                    "changing_lines": []
                },
                "needs_follow_up": False,
                "created_at": "2026-02-13T12:00:00"
            }
        }


class CreateDivinationRequest(BaseModel):
    """创建占卜请求"""
    user_id: str = Field(..., description="用户ID")
    question: str = Field(..., min_length=1, max_length=500, description="占卜问题")
    event_type: Optional[str] = Field(None, description="事件类型")
    version: str = Field("CN", description="版本（CN/TAROT）")
    orientation: Optional[str] = Field(None, description="方位")
    spread: Optional[str] = Field(None, description="牌阵类型")
    intent: Optional[str] = Field(None, description="意图类型")
    context: Optional[Dict[str, Any]] = Field(None, description="上下文信息")


class DivinationSession(BaseModel):
    """占卜会话"""
    id: str
    user_id: str
    version: str
    question: str
    event_type: Optional[str] = None
    orientation: Optional[str] = None
    spread: Optional[str] = None
    status: str
    follow_up_count: int = 0
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class DivinationTaskAccepted(BaseModel):
    """占卜任务已受理响应"""
    accepted: bool = Field(True, description="是否已受理")
    session_id: str = Field(..., description="占卜会话 ID")
    status: str = Field("processing", description="当前状态")
    status_url: str = Field(..., description="查询状态/结果的 API 路径")
    message: str = Field("占卜任务已受理，正在处理中", description="提示信息")
    created_at: datetime
